packages/js/product-editor/src/components/block-template-provider/block-template-provider.tsx

```diff
@@ -0,0 +1,84 @@
+/**
+ * External dependencies
+ */
+import { createElement, useLayoutEffect, useState } from '@wordpress/element';
+import {
+	BlockInstance,
+	synchronizeBlocksWithTemplate,
+} from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { getVisibleBlocks } from './get-visible-blocks';
+import { Provider } from './context';
+import { Template } from './types';
+import { TestTemplateComponent } from './test-template-component';
+
+type BlockTemplateProviderProps = {
+	children: JSX.Element;
+	initialTemplate: string;
+	onChange: (
+		blocks: BlockInstance[],
+		options: {
+			[ key: string ]: unknown;
+		}
+	) => void;
+	templates: Template[];
+};
+
+export function BlockTemplateProvider( {
+	children,
+	initialTemplate,
+	onChange,
+	templates,
+}: BlockTemplateProviderProps ) {
+	const [ selectedTemplate, setSelectedTemplate ] =
+		useState( initialTemplate );
+	const [ hiddenBlocks, setHiddenBlocks ] = useState< string[] >( [] );
```

Comment:
I would suggest moving this state to a redux store, instead of keeping it in a wrapped provider.
I think the using a store would be quite a bit more performant for something like this, and it also follow more inline with Gutenberg, given they store most of the editor state in a store and the very basic things that don't change to often within a provider.
Similar to the `core/edit-site` data store.

Reply:
I think pushing this to a data store could be a good idea, though I'll note that we still need the wrapped provider for this to work and the two aren't mutually exclusive.

For example, the entity provider acts as the provider with a set of exported functions that can be used, though the underlying data is stored in a data store.

https://github.com/WordPress/gutenberg/blob/14600f31582ba0ebe7343130d864ac6dba1d73db/packages/core-data/src/entity-provider.js#L106-L136

If we do choose to keep this in a data store, we'll need to make sure that the data store keeps track of changes on the template level (probably by ID) instead of globally, since we don't want one template interfering with another.

Reply:
> For example, the entity provider acts as the provider with a set of exported functions that can be used, though the underlying data is stored in a data store.

In this case the entity provider only value it stores is the entity `id`, it makes use of that (+ the passed in entity type and kind ) to decipher everything from the data store.
The provider itself doesn't expose any functions or additional data, this all comes from the data store based on the `id`.
I think we can do something similar, and if we think `templateID` is needed on top of the product id, we could only expose that through the provider.